### PR TITLE
Fixed #26056 -- ArrayField does not work with ValueListQuerySet.

### PR DESCRIPTION
**Pull request by atombrella**

### CODE REVIEW

File: django/contrib/postgres/fields/array.py [hunk 1]
```diff
@@ -219,6 +220,11 @@ def as_sql(self, qn, connection):
         sql = '%s::%s' % (sql, self.lhs.output_field.db_type(connection))
         return sql, params
 
+    def process_rhs(self, compiler, connection):
+        if isinstance(self.rhs, Query):
+            return 'ARRAY(%s)' % self.rhs, []
```
Comment (charettes) [inline django/contrib/postgres/fields/array.py]: You shouldn't use the string representation of the query but someting along the lines of what `django.db.models.fields.related.related_lookups.RelatedIn` does.

File: tests/postgres_tests/test_array.py [hunk 2]
```diff
@@ -235,6 +235,15 @@ def test_overlap_charfield(self):
...
         )
 
+    def test_overlap_values(self):
+        # 26056
+        instance = CharArrayModel(field=['hello', 'goodbye'])
+        instance.save()
+        qs = CharArrayModel.objects.filter(field__overlap=CharArrayModel.objects.values('field'))
+        self.assertQuerysetEqual(qs, [
+            (['hello', 'goodbye']),
+        ], transform=lambda row: row.field)
```
Comment (charettes) [inline tests/postgres_tests/test_array.py]: `self.assertSequenceEqual(qs.values_list('field', flat=True), [(['hello', 'goodbye'])])`

File: django/contrib/postgres/fields/array.py (hunk 1 above)
Comment (atombrella) [inline django/contrib/postgres/fields/array.py]: Good. I'll try to figure this out.

**Comment by timgraham** [comment]
Is "Patch needs improvement" on the ticket the correct status? There's a recent comment asking if `Subquery` could be used for this.

**Comment by atombrella** [comment]
I think I hit a wall, so perhaps this could be closed? If somebody else wishes to continue my efforts, please do so. Using queries is a consideration in other tickets (Coalesce) as well, which would be probably overlap with this.

**Comment by timgraham** [comment]
Okay, thanks.

**Comment by kosz85** [comment]
It's easy to fix with Subquery, but Subquery isn't enough:
```
from django.db.models import Func


class Array(Func):
    function='array'
```
Then you can do:
```
Model.objects.filter(some_field__overlap=Array(Subquery(OtherModel.objects.all().values_list('id', flat=True)))
````
And that works just fine.

**Pull request closed without merging**